include/envoy/server/instance.h

```diff
@@ -152,6 +152,11 @@ class Instance {
    */
   virtual void shutdown() PURE;
 
+  /**
+   * @return whether the shutdown method has been called.
+   */
+  virtual bool isShutdown() PURE;
```

Comment:
Do you know if the "don't call virtual functions in destructor" (e.g. https://www.artima.com/cppsource/nevercall.html) rule applies here?

Reply:
Thanks for the pointer. I think the way I'm using it here is safe though - here is the relevant bit "Once a derived class destructor has run, the object's derived class data members assume undefined values, so C++ treats them as if they no longer exist. Upon entry to the base class destructor, the object becomes a base class object".

In this case, we're calling the virtual method (isShutdown) from the derived class destructor (InstanceImpl) not that of the base class (Instance) so the object should still be treated as an InstanceImpl and call the correct method.

Reply:
@htuch Actually `clang-analyzer-optin.cplusplus.VirtualCall` check will warn about this (though, not an error now so it will pass) in clang-tidy CI. It doesn't seems the case?

> In this case, we're calling the virtual method (isShutdown) from the derived class destructor (InstanceImpl) not that of the base class (Instance) so the object should still be treated as an InstanceImpl and call the correct method.

@eziskind if you make the virtual method called by the derived class destructor `final`, then clang-analyzer won't complain.

Reply:
done